### PR TITLE
Brad 2020

### PR DESCRIPTION
**Pull request by ac0zj**
Hi David, here's the pull request.  Brad

### CODE REVIEW

File: src/fdmdv2_main.cpp [hunk 1]
```diff
@@ -1111,7 +1116,7 @@ void MainFrame::OnTimer(wxTimerEvent &evt)
 	m_textSync->SetLabel("Modem");
      }
     g_prev_State = g_State;
-    if (g_mode == FREEDV_MODE_700D) {
+    if ((g_mode == FREEDV_MODE_700D) || (g_mode == FREEDV_MODE_2020)){
```
Comment (drowe67) [inline src/fdmdv2_main.cpp]: Good thinking! Hhowever the interleaver (at least past one frame) seems really used, so I doubt it will be used for 2020

Comment (srsampson) [reply src/fdmdv2_main.cpp]: The interleaver might be good for broadcasting.

File: src/fdmdv2_main.cpp [hunk 2]
```diff
@@ -2602,18 +2616,24 @@ void MainFrame::OnTogBtnOnOff(wxCommandEvent& event)
...
+            struct CODEC2 *c2 = freedv_get_codec2(g_pfreedv);
+            if (c2 != NULL) {
+                codec2_set_lpc_post_filter(c2,
+                                           wxGetApp().m_codec2LPCPostFilterEnable,
+                                           wxGetApp().m_codec2LPCPostFilterBassBoost,
+                                           wxGetApp().m_codec2LPCPostFilterBeta,
+                                           wxGetApp().m_codec2LPCPostFilterGamma);
+            }
+            
             // Init Speex pre-processor states
```
Comment (drowe67) [inline src/fdmdv2_main.cpp]: @ac0zj How about this Brad?  I thought this was perhaps easier than adding (yet another) API function.

Comment (ac0zj) [reply src/fdmdv2_main.cpp]: I think that's fine - simple and makes sense!

**Comment by drowe67** [comment]
@ac0zj or @hobbes1069  could you pls:
1. Update READMe.md on how to build freedv-gui with 2020 support.  I am not sure what cmake line to use
1. Alternatively get .travis.yml working in this branch, as that will show me how to do it.

**Comment by hobbes1069** [comment]
> @ac0zj or @hobbes1069 could you pls:
> 
> 1. Update README.md on how to build freedv-gui with 2020 support for Linux.  I am not sure what cmake line to use.
> 2. Alternatively get .travis.yml working in this branch, as that will show me how to do it.

Ok, I want to be carefull... This is Pull #1 but mine is Pull #9. Do I need to merge my pull request and then fix Travis CI?

**Comment by drowe67** [comment]
@hobbes1069 no I think that would be too big a step right now.  I just need the build instructions pls so I can start testing FreeDV 2020 ASAP please.

**Comment by hobbes1069** [comment]
Ok @drowe67 I think we're talking around each other. All of my work to make things build has been done under rms-2020 which was branched from brad-2020. I can't do anything to make this branch build as we discussed creating my own branch instead of updating this one. Right now I'm working on the dependency "gatherer" for the windows package because it's not picking up the jpeg dependency. I'm going two levels down but it appears that's not enough so I'm work on a recursive strategy but I'm beating my head against the wall to get the logic right without creating an infinite loop.

**Comment by drowe67** [comment]
OK no problem @hobbes1069 I'm sure Brad can tell me how to make this branch build.  Let's continue the discussion re https://github.com/drowe67/freedv-gui/pull/9 in https://github.com/drowe67/freedv-gui/pull/9.

**Comment by hobbes1069** [comment]
We can continue with the manual double build using travis or try to use my "USE_INTERNAL_CODEC2" option to let cmake handle it. I need to double check the Build<proj>.cmake files to make sure they reference the correct branch now.

**Comment by hobbes1069** [comment]
I'm at work and need to take care of a few things. You need to add -DLPCNET_BUILD_DIR to travis and it should work.

**Comment by drowe67** [comment]
@ac0zj I have a TODO list for 2020 here https://github.com/drowe67/freedv-gui/issues/11

I'm running on over the air on the bench between a IC7200 and FT817.  The IC7200 is playing a canned modulated file, and 817 is connected to a Linux machine running freedv-gui, output through small loudspeakers  I'm getting a few "I can't believe this is HF" moments .....

The output level seems a bit high on the "to speaker" tab, and I can maybe hear some clipping in the output audio.  I recall 700D needed some audio level boosting and may still have that somewhere in the 2020 code.

File: build_windows.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Comment (drowe67) [inline build_windows.sh]: @hobbes1069 @ac0zj I wrote this script to automate the windows X-compile.  Does this work for you guys?  If so, can I modify the README.md to use this script?

**Comment by drowe67** [comment]
@ac0zj I've been working up the Windows side and have the problem traced to the NN decoder https://github.com/drowe67/LPCNet/pull/7#issuecomment-491425407

I have a few local Hams set up to try the Linux version this weekend.  I'd like to see how the "human factors" play out in a PTT QSO, in particular sync time.

**Comment by ac0zj** [comment]
David,

Sounds good.  I got a bit sidetracked with another radio-type project.
I have a co-worker in New Jersey who is interested in FreeDV, so I'll see
if I can get him set up.  Otherwise I can try a contact with other U.S.
FreeDV followers.

Cheers,
Brad

File: src/fdmdv2_main.cpp [hunk 4]
```diff
@@ -3764,50 +3938,48 @@ void txRxProcessing()
...
+ 	if (g_dump_fifo_state) {
+	  // If this drops to zero we have a problem as we will run out of output samples
+	  // to send to the sound driver via PortAudio
+	  fprintf(stderr, "outfifo1 used: %6d free: %6d nsam_one_modem_frame: %d\n",
+                  codec2_fifo_used(cbData->outfifo1), codec2_fifo_free(cbData->outfifo1), nsam_one_modem_frame);
+	}
 
+        int nsam_in_48 = g_soundCard2SampleRate * freedv_get_n_speech_samples(g_pfreedv)/freedv_get_speech_sample_rate(g_pfreedv);
+        assert(nsam_in_48 < 10*N48);
         while((unsigned)codec2_fifo_free(cbData->outfifo1) >= nsam_one_modem_frame) {
```
Comment (drowe67) [inline src/fdmdv2_main.cpp]: @ac0zj I think this should be ```while((unsigned)codec2_fifo_free(cbData->outfifo1) >= nsam_one_modem_frame)``` to ensure output fifo stays full?

Perhaps we can now try smaller FIFO sizes?

**Comment by ac0zj** [comment]
I agree - ugh, it should be " >= ", not "<".  Yes, I bet we can go back to
smaller FIFO.  And I bet it will work well!!

**Pull request merged**